src/System.Buffers/src/System/Buffers/DefaultArrayPool.cs

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Buffers
+{
+    internal sealed class DefaultArrayPool<T> : ArrayPool<T>
+    {
+        private const int MinimiumArraySize = 16;
+        private int _maxArrayLength;
+        private int _numberOfArraysPerBucket;
+        private bool _isTObjectType;    // cache this to only do the reflection call once
```

Comment:
This doesn't need to be stored per instance, it can be stored per type, e.g. a static
```C#
private static readonly bool s_isTObjectType = GetIsTObjectType();
```
That said, the current definition isn't correct, or more specifically it's not providing the behavior I think you're hoping for.  It looks like you're using this to determine whether to always clear the array in order to avoid leaks, but even if T is a value type, it can still contain reference fields, e.g. your code would clear in the case of:
```C#
ArrayPool<Object> ...;
```
but not in the case of:
```C#
struct ObjectWrapper
{
     public Object Data;
}

ArrayPool<ObjectWrapper> ...;
```
yet they have the same issue.

See https://github.com/dotnet/coreclr/issues/1135.